### PR TITLE
UAF-2967 Procurement Card Reconciliation Step 4 (Postback)

### PR DESCRIPTION
**Pull request by mmoen**
Modified ProcurementCardCreateDocumentServiceImpl.java to set the description the same way it is done in KFS 3.0. This fixes the issue where backpost is allowed and the description is changed to add additional text to the beginning of the descripition. When the GLPE's were saved, transactionLedgerEntryDescription was over the maximum 40 characters the database allows for that field.
Modified ProcurementCardDocument.java to override the customizeExplicitGeneralLedgerPendingEntry method to include backposting, plus helper methods.

### CODE REVIEW

File: kfs-core/src/main/java/edu/arizona/kfs/fp/batch/service/impl/ProcurementCardCreateDocumentServiceImpl.java [hunk 1]
```diff
@@ -398,6 +399,25 @@ public boolean autoApproveProcurementCardDocuments() {
         return true;
     }
     
+    @Override
+    protected String createAndValidateAccountingLines(org.kuali.kfs.fp.document.ProcurementCardDocument pcardDocument, ProcurementCardTransaction transaction, ProcurementCardTransactionDetail docTransactionDetail) {
+        //get default values from custom Procurement Cardholder table
+        ProcurementCardDefault procurementCardHolderDetail = getProcurementCardDefault(transaction.getTransactionCreditCardNumber());
+
+        String cardHolderName = transaction.getCardHolderName();
+        if (cardHolderName.length() > 24) {
```
Comment (nataliac06) [inline kfs-core/src/main/java/edu/arizona/kfs/fp/batch/service/impl/ProcurementCardCreateDocumentServiceImpl.java]: Please add a constant in the PcardDocument for the cardHolderNameMaxLength and replace the hard-coded 24 value.

File: kfs-core/src/main/java/edu/arizona/kfs/fp/document/ProcurementCardDocument.java [hunk 2]
```diff
@@ -32,6 +43,55 @@ public ProcurementCardHolder getProcurementCardHolder() {
...
+    public void customizeExplicitGeneralLedgerPendingEntry(GeneralLedgerPendingEntrySourceDetail postable, GeneralLedgerPendingEntry explicitEntry) {
+        Date temp = getProcurementCardTransactionPostingDetailDate();
+        
+        if( temp != null && allowBackpost(temp) ) {
+            Integer prevFiscYr = getPreviousFiscalYear();
+            
+            explicitEntry.setUniversityFiscalPeriodCode(FINAL_ACCOUNTING_PERIOD);
+            explicitEntry.setUniversityFiscalYear(prevFiscYr);
+            
+            if( !getDocumentHeader().getDocumentDescription().contains("FY " + prevFiscYr) ) {
```
Comment (nataliac06) [inline kfs-core/src/main/java/edu/arizona/kfs/fp/document/ProcurementCardDocument.java]: Please refactor the if statement:
String documentDescription = getDocumentHeader().getDocumentDescription();
if ( StringUtils.isNotEmpty(documentDescription) && documentDescription.contains("FY ....){
    getDocumentHeader().setDocumentDescription("FY "+prevFiscYr+" " + documentDescription);
}
Do we have an idea regarding what to do if the documentDescription is empty? Is there a default value we should set it to?

File: kfs-core/src/main/java/edu/arizona/kfs/fp/document/ProcurementCardDocument.java [hunk 3]
```diff
@@ -93,4 +153,36 @@ protected String buildDocumentTitle(String title) {
...
         return title;
     }
-}
\ No newline at end of file
+    
+    protected boolean allowBackpost(Date tranDate) {
+        ParameterService parameterService = SpringContext.getBean(ParameterService.class);
+        UniversityDateService universityDateService = SpringContext.getBean(UniversityDateService.class);
+       
+        int allowBackpost = (Integer.parseInt(parameterService.getParameterValueAsString(ProcurementCardLoadStep.class, ALLOW_BACKPOST_DAYS)));
```
Comment (nataliac06) [inline kfs-core/src/main/java/edu/arizona/kfs/fp/document/ProcurementCardDocument.java]: Please separate this in two steps:
String allowBackpostParam = parameterService.getParameter.....
int allowBackpost = 0; -> also please check if this is an OK default value  
if (StringUtils.isNotEmpty() && StringUtils.isNumeric( allowBackPostDyas ) ){
    allowBackpost = Integer.parseInt( allowBackpostParam );
}

**Review by nataliac06 — CHANGES_REQUESTED** [review]

File: kfs-core/src/main/java/edu/arizona/kfs/fp/batch/service/impl/ProcurementCardCreateDocumentServiceImpl.java (hunk 1 above)
Comment (mmoen) [reply kfs-core/src/main/java/edu/arizona/kfs/fp/batch/service/impl/ProcurementCardCreateDocumentServiceImpl.java]: fixed

File: kfs-core/src/main/java/edu/arizona/kfs/fp/document/ProcurementCardDocument.java (hunk 2 above)
Comment (mmoen) [reply kfs-core/src/main/java/edu/arizona/kfs/fp/document/ProcurementCardDocument.java]: Fixed. As for the documentDescription, Misty said the value is  the credit card last four digits and card name, which is being set in the createAndValidateAccountingLines method of (edu/arizona)ProcurementCardCreateDocumentServiceImpl.java.

**Comment by nataliac06** [comment]
Looks good.

**Pull request merged**